packages/vkui/src/components/BaseGallery/CarouselBase/CarouselBase.tsx

```diff
@@ -0,0 +1,561 @@
+    React.useState<ControlElementsState>(CONTROL_ELEMENTS_STATE);
+
+  const { window } = useDOM();
+  const hasPointer = useAdaptivityHasPointer();
+
+  const isCenterWithCustomWidth = slideWidth === 'custom' && align === 'center';
+
+  const validateIndent = (value: number) => {
+    const localMax = layoutState.current.max ?? 0;
+    const localMin = layoutState.current.min ?? 0;
+
+    if (value < localMin) {
+      return localMin;
+    } else if (value > localMax) {
+      return localMax;
+    }
+
+    return value;
+  };
+
+  /*
+   * Считает отступ слоя галереи
+   */
+  const calculateIndent = (targetIndex: number) => {
+    if (layoutState.current.isFullyVisible) {
+      return 0;
+    }
+
+    const targetSlide = layoutState.current.slides?.length
+      ? layoutState.current.slides[targetIndex]
+      : null;
+
+    if (targetSlide) {
+      const { coordX, width } = targetSlide;
+
+      if (isCenterWithCustomWidth) {
+        const viewportWidth = layoutState.current.viewportOffsetWidth ?? 0;
+        return viewportWidth / 2 - coordX - width / 2;
+      }
+
+      return -1 * coordX;
+    }
+
+    return 0;
+  };
+
+  const transformCssStyles = (shiftX: number, animation = false) => {
+    shiftState.current.loopPoints.forEach((loopPoint) => {
+      const { target, index } = loopPoint;
+      const slide = slidesStore.current[`slide-${index}`];
+      if (slide) {
+        slide.style.transform = `translate3d(${target(shiftX)}px, 0, 0)`;
+      }
+    });
+
+    if (layerRef.current) {
+      layerRef.current.style.transform = `translate3d(${shiftX}px, 0, 0)`;
+      layerRef.current.style.transition = animation
+        ? `transform ${ANIMATION_DURATION}s cubic-bezier(.1, 0, .25, 1)`
+        : '';
+    }
+  };
+
+  const requestTransform = (shiftX: number, animation = false) => {
+    if (animationFrameRef.current !== null) {
+      cancelAnimationFrame(animationFrameRef.current);
+    }
+    const { snaps } = shiftState.current;
+
+    animationFrameRef.current = requestAnimationFrame(() => {
+      if (shiftX > snaps[0]) {
+        shiftXCurrentRef.current = -shiftState.current.contentSize + snaps[0];
+        shiftX = shiftXCurrentRef.current + shiftXDeltaRef.current;
+      }
+      const lastPoint =
+        layoutState.current.slides[layoutState.current.slides.length - 1].width +
+        layoutState.current.slides[layoutState.current.slides.length - 1].coordX;
+
+      if (shiftX <= -lastPoint) {
+        shiftXCurrentRef.current = Math.abs(shiftXDeltaRef.current) + snaps[0];
+      }
+      transformCssStyles(shiftX, animation);
+    });
+  };
+
+  const initializeSlides = (options: { animation?: boolean } = {}) => {
+    let localSlides =
+      React.Children.map(children, (_item: React.ReactNode, i: number): GallerySlidesState => {
+        const elem = slidesStore.current[`slide-${i}`];
+        return {
+          coordX: elem?.offsetLeft ?? 0,
+          width: elem?.offsetWidth ?? 0,
+        };
```

Comment:
```suggestion
        const elem = slidesStore.current[`slide-${i}`] || { offsetLeft: 0, offsetWidth: 0 };
        return {
          coordX: elem.offsetLeft,
          width: elem.offsetWidth,
        };
```
можно вот так оптимизировать

optional chaining не дешевый с учётом того, во что компиляторы трансформируют